### PR TITLE
Add a keyboard shortcut to save changes in the site editor

### PR DESCRIPTION
**Pull request by youknowriad**
I found myself hitting ctrl+s  to save the changes in the site editor but it didn't work.
This PR implements that shortcut for the site editor (to match the post editor)

### CODE REVIEW

**Comment by ntsekouras** [comment]
This seems a little risky IMO to save when you might have changed multiple entities (site, template/template parts etc..). On one hand is something helpful when you know exactly what you've changed, on the other hand you might end up with changes you wouldn't want to. 

I personally like that you can review the changes (affected entities) and at least for now, I usually check them.

I'm not against it but I think it can lead to confusion (in some users) in cases like changing the `Site Title block` in a template part (ex header).

**Comment by youknowriad** [comment]
Good point Nik. That said, I feel someone familiar with the shortcut is probably an advanced user who knows what he's doing.  Also right now, if you click the button, reviewing changes is optional (only if you open the panel to check). What do you think?

**Comment by ntsekouras** [comment]
Yeah, as I said I'm not against it. Since FSE is relatively new, for now it's definitely safe. We can add this and revisit if any issues are raised.

>familiar with the shortcut is probably an advanced user 
Though `Ctrl+S` might be the most common shortcut ever :)

File: packages/edit-site/src/components/save-button/index.js [hunk 1]
```diff
@@ -6,26 +6,60 @@ import { some } from 'lodash';
...
+			} = select( 'core' );
+			const dirtyEntityRecords = __experimentalGetDirtyEntityRecords();
+			return {
+				isDirty: dirtyEntityRecords.length > 0,
+				isSaving: some( dirtyEntityRecords, ( record ) =>
+					isSavingEntityRecord( record.kind, record.name, record.key )
+				),
+				getDirtyEntityRecords: __experimentalGetDirtyEntityRecords,
+			};
+		}
```
Comment (ntsekouras) [inline packages/edit-site/src/components/save-button/index.js:31]: We should add an empty array for dependencies here, no?

File: packages/edit-site/src/components/save-button/index.js [hunk 2]
```diff
@@ -6,26 +6,60 @@ import { some } from 'lodash';
...
+	useEffect( () => {
+		registerShortcut( {
+			name: 'core/edit-site/save',
+			category: 'global',
+			description: __( 'Save your changes.' ),
+			keyCombination: {
+				modifier: 'primary',
+				character: 's',
+			},
+		} );
```
Comment (ntsekouras) [inline packages/edit-site/src/components/save-button/index.js:44]: Same comment as above for dependencies.

**Review by hc0503 — APPROVED** [review]

**Comment by jasmussen** [comment]
Yep, seems like an obvious win.

Question though — does this save mulitiple entities? Or just the post?

**Comment by youknowriad** [comment]
It saves everything

**Comment by jasmussen** [comment]
By that you mean it saves everything, _and_ publishes those changes? I.e. when you save a change to a template part, it's effectively live, correct?

While I'd be happy to try this for consistency, I do share some nervousness about this in the long term. Can we limit it to save only post content, or the area that has focus?

**Comment by youknowriad** [comment]
> While I'd be happy to try this for consistency, I do share some nervousness about this in the long term. Can we limit it to save only post content, or the area that has focus?

I think saving only "post content" doesn't make sense because you rarely update post content on the site editor, you work on templates instead. 

Only save the area that has focus is very ambigouis, say you add a site title block, you tweak it and modify the title. do you save the title of the site or the template where the block was added/modified?

**Comment by jasmussen** [comment]
> I think saving only "post content" doesn't make sense because you rarely update post content on the site editor, you work on templates instead.

Solid point. But in thinking about that, there isn't _actually_ a save button at all, on the site editor — so ⌘S is a little misleading. There's only a "Update Design", button, and by immediately saving, we're bypassing this screen:

Screenshot 2020-11-12 at 11 39 56

Since the very beginning, I think explicit "Save" buttons are a vestige of the past, and that autosave is the future. Figma showing a snackbar informing you that you don't have to save, is ultimately where I'd like us to be. So by that logic, ⌘S should not do anything on this screen at all.

However could we open the pre-publish dialog when you ⌘S in the site editor? I don't think it'd be the correct thing to do, but it might be worth just getting a feel for it?

**Comment by youknowriad** [comment]
> Figma showing a snackbar informing you that you don't have to save, is ultimately where I'd like us to be.

This actually validates this PR for me because I suppose auto-saving means auto-saving everything.

> However could we open the pre-publish dialog when you ⌘S in the site editor? I don't think it'd be the correct thing to do, but it might be worth just getting a feel for it?

Yes, we can try that. 

I'd like to clarify that this PR is based on my own frustration working on that screen and having to go through the "save design" panel over and over again for each change.

**Comment by jasmussen** [comment]
> This actually validates this PR for me because I suppose auto-saving means auto-saving everything.

Another good point. But saving a site title change, to me, intrinsically feels more like a _publishing act_ than something you'd ever have _auto-saved_. In that light, I see ⌘S as a speeding up of the auto-save, and that we just don't have any auto-save for the site editor at all.

**Comment by MichaelArestad** [comment]
> Another good point. But saving a site title change, to me, intrinsically feels more like a publishing act than something you'd ever have auto-saved. In that light, I see ⌘S as a speeding up of the auto-save, and that we just don't have any auto-save for the site editor at all.

@jasmussen Right. Really, we need to be able to save everything without publishing everything. Exactly how saving works on a post. The Publish flow, like Posts, should be done through the Publish (in this case Multi-entity saving) flow. This way, I can edit the site and my post, save the changes, and choose which changes to publish on an individual basis if needed without accidentally publishing a change post/page content.

**Comment by Addison-Stavlo** [comment]
What if ⌘S opens the save panel if it is not open, and acts to save all selected in the panel if it is open?  So for your flow @youknowriad you would need to ⌘S twice in a row to get your result.

This would help keep some consistency with having a ⌘S shortcut with the post editor, but still allow the implications of multiple entities to be more specific.

> I feel someone familiar with the shortcut is probably an advanced user who knows what he's doing.

Anyone may be in the habit of using ⌘S from the post editor, and immediately try that as their first impulse for saving in the site editor.  Integrating this with the save panel would safeguard things quite a bit. 🤔

**Comment by youknowriad** [comment]
Closing this for now.

**Pull request closed without merging**